compression_verifier.go

```diff
@@ -0,0 +1,238 @@
+package ghostferry
+
+import (
+	"crypto/md5"
+	"database/sql"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/golang/snappy"
+	"github.com/siddontang/go-mysql/schema"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// CompressionSnappy is used to identify Snappy (https://google.github.io/snappy/) compressed column data
+	CompressionSnappy = "SNAPPY"
+)
+
+type (
+	columnCompressionConfig map[string]string
+
+	// TableColumnCompressionConfig represents compression configuration for a
+	// column in a table as table -> column -> compression-type
+	// ex: books -> contents -> snappy
+	TableColumnCompressionConfig map[string]columnCompressionConfig
+)
+
+// UnsupportedCompressionError is used to identify errors resulting
+// from attempting to decompress unsupported algorithms
+type UnsupportedCompressionError struct {
+	table     string
+	column    string
+	algorithm string
+}
+
+func (e UnsupportedCompressionError) Error() string {
+	return "Compression algorithm: " + e.algorithm +
+		" not supported on table: " + e.table +
+		" for column: " + e.column
+}
+
+// CompressionVerifier provides support for verifying the payload of compressed columns that
+// may have different hashes for the same data by first decompressing the compressed
+// data before fingerprinting
+type CompressionVerifier struct {
+	logger *logrus.Entry
+
+	// supportedAlgorithms provide O(1) lookup to check if a configured algorithm is supported
+	supportedAlgorithms map[string]struct{}
+
+	// tableColumnCompressions is represented as table -> column -> compression-type
+	tableColumnCompressions TableColumnCompressionConfig
```

Comment:
I don't understand why `tableColumnCompressions` is a field on `CompressionVerifier`. It's never accessed from within the methods here. It's in the constructor but that's just to populate the field – not consume it

The only other time I see it being accessed is from within the `IterativeVerifier`. Is that a sign that it should not be here? Perhaps it should just be a config on the `IterativeVerifier` directly?

Reply:
It's currently part of the `CompressionVerifier` because it's the configuration for it. Given the `CompressionVerifier` is also a component of the `IterativeVerifier` it's essentially part of the `IterativeVerifier` already, but contained within the `CompressionVerifier` as the `CompressionVerifier` contains the logic and info for all things compressed. It's also not accessed within the `CompressionVerifier`, but referenced at runtime and then passed to the `GetCompressedHashes` method as that's how I understood your earlier comment about the direct logic flow: https://github.com/Shopify/ghostferry/pull/44#pullrequestreview-140863219. 

Does that make sense? If you feel we should still move it let me know so we can discuss because I want this to make sense 👍